File: java/src/main/java/algorithms/arrays/ReverseString.java
package algorithms.arrays;

/**
 * LeetCode #344 - Reverse String
 * Difficulty: Easy
 * 
 * Reverses a character array in-place using O(1) extra space.
 * Uses two-pointer technique.
 * 
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */
public class ReverseString {
    
    /**
     * Reverses the input character array in-place.
     * 
     * @param s the character array to reverse
     */
    public void reverseString(char[] s) {
        int left = 0;
        int right = s.length - 1;
        
        while (left < right) {
            // Swap characters at left and right pointers
            char temp = s[left];
            s[left] = s[right];
            s[right] = temp;
            
            // Move pointers inward
            left++;
            right--;
        }
    }
}

File: elixir/lib/algorithms/arrays/reverse_string.ex
defmodule Algorithms.Arrays.ReverseString do
  @moduledoc """
  LeetCode #344 - Reverse String
  Difficulty: Easy

  Reverses a list of characters.

  Time Complexity: O(n)
  Space Complexity: O(n) - Elixir uses immutable data structures
  """

  @doc """
  Reverses the input list of characters.

  ## Examples

      iex> Algorithms.Arrays.ReverseString.reverse_string(["h", "e", "l", "l", "o"])
      ["o", "l", "l", "e", "h"]

  """
  @spec reverse_string([String.t()]) :: [String.t()]
  def reverse_string(s) do
    Enum.reverse(s)
  end

  @doc """
  Alternative implementation using reduce for educational purposes.
  """
  @spec reverse_string_manual([String.t()]) :: [String.t()]
  def reverse_string_manual(s) do
    Enum.reduce(s, [], fn char, acc -> [char | acc] end)
  end
end
